File: Jack.Business/Presenca.vb
Public Class Presenca

    Implements ICrud(Of Model.Presenca, Integer)

    Public Sub New()

    End Sub

    Public Function Delete(oTipo As Model.Presenca) As Boolean Implements ICrud(Of Model.Presenca, Integer).Delete

        Dim oDados As Data.Presenca = Nothing
        Dim blnRetorno As Boolean = False

        Try
            oDados = New Data.Presenca
            blnRetorno = oDados.Delete(oTipo)
        Catch ex As Exception
            blnRetorno = False
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return blnRetorno

    End Function

    Public Function Find(Identifier As Integer) As Model.Presenca Implements ICrud(Of Model.Presenca, Integer).Find

        Dim oDados As Data.Presenca = Nothing
        Dim lstRetorno As IList(Of Model.Presenca) = Nothing

        Try
            oDados = New Data.Presenca
            lstRetorno = oDados.LoadAll()
        Catch ex As Exception
            lstRetorno = Nothing
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return lstRetorno

    End Function

    Public Function Insert(oTipo As Model.Presenca) As Boolean Implements ICrud(Of Model.Presenca, Integer).Insert

        Dim oDados As Data.Presenca = Nothing
        Dim blnRetorno As Boolean = False

        Try
            oDados = New Data.Presenca
            blnRetorno = oDados.Insert(oTipo)
        Catch ex As Exception
            blnRetorno = False
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return blnRetorno
    End Function

    Public Function InsertLote(lstFamilia As IList(Of Model.Presenca)) As Boolean

        Dim blnRetorno As Boolean = False
        Dim intCont As Integer = 0
        Try
            For Each oDado As Model.Presenca In lstFamilia
                blnRetorno = Insert(oDado)
                If Not blnRetorno Then
                    Throw New Exception("Tentativa de Gravar Presença falhou. Antes foram gravados " + intCont.ToString + " de " + lstFamilia.Count + " registros.")
                End If
                intCont = intCont + 1
            Next
        Catch ex As Exception
            blnRetorno = False
            Throw ex
        End Try

        Return blnRetorno
    End Function

    Public Function LoadAll() As IList(Of Model.Presenca) Implements ICrud(Of Model.Presenca, Integer).LoadAll

        Dim oDados As Data.Presenca = Nothing
        Dim lstRetorno As IList(Of Model.Presenca) = Nothing

        Try
            oDados = New Data.Presenca
            lstRetorno = oDados.LoadAll()
        Catch ex As Exception
            lstRetorno = Nothing
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return lstRetorno

    End Function

    Function Load(intReuniao As Integer) As IList(Of Model.Familia)

        Dim oDados As Data.Presenca = Nothing
        Dim lstRetorno As IList(Of Model.Familia) = Nothing

        Try
            oDados = New Data.Presenca
            lstRetorno = oDados.Load(intReuniao)
        Catch ex As Exception
            lstRetorno = Nothing
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return lstRetorno

    End Function

    Public Function Update(oTipo As Model.Presenca) As Boolean Implements ICrud(Of Model.Presenca, Integer).Update

        Dim oDados As Data.Presenca = Nothing
        Dim blnRetorno As Boolean = False

        Try
            oDados = New Data.Presenca
            blnRetorno = oDados.Update(oTipo)
        Catch ex As Exception
            blnRetorno = False
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return blnRetorno
    End Function

    Public Function ObterPresencaPorMae(intFamilia As Integer, intAno As Integer) As List(Of Model.FamiliaPresenca)

        Dim oDados As Data.Presenca = Nothing
        Dim lstRetorno As List(Of Model.FamiliaPresenca) = Nothing

        Try
            oDados = New Data.Presenca
            lstRetorno = oDados.ObterPresencaPorMae(intFamilia, intAno)
        Catch ex As Exception
            lstRetorno = Nothing
            Throw ex
        Finally
            oDados = Nothing
        End Try

        Return lstRetorno

    End Function
End Class

File: Jack.Business/Presenca.cs
using Jack.Model.DTOs;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Jack.Business
{

    public class Presenca : ICrud<Model.Presenca, int>
    {


        private readonly Data.Familia repFamilia;
        private readonly Data.Reuniao repReuniao;
        private readonly Data.Presenca repPresenca;
        private readonly Data.IUnitWork unidadeTrabalho;

        public Presenca()
        {
            unidadeTrabalho = new Data.UnitWork();
            repReuniao = new Data.Reuniao(unidadeTrabalho);
            repFamilia = new Data.Familia(unidadeTrabalho);
            repPresenca = new Data.Presenca(unidadeTrabalho);

        }

        public bool Delete(Model.Presenca oTipo)
        {

            bool blnRetorno = false;

            try
            {
                blnRetorno = repPresenca.Delete(oTipo);
            }
            catch (Exception ex)
            {
                blnRetorno = false;
                throw ex;
            }

            return blnRetorno;

        }

        public Model.Presenca Find(int Identifier)
        {

            Model.Presenca oRetorno = null;

            try
            {
                oRetorno = repPresenca.Find(Identifier);
            }
            catch (Exception ex)
            {
                oRetorno = null;
                throw ex;
            }

            return oRetorno;

        }
        public bool Registrar(DTOPresenca presencaMae)
        {
            bool blnRetorno = false;
            Model.Presenca modelPresenca = null;

            try
            {
                modelPresenca = new Model.Presenca();

                modelPresenca.Familia = repFamilia.Find(presencaMae.Familia);
                modelPresenca.Reuniao = repReuniao.Find(presencaMae.Reuniao);
                blnRetorno = repPresenca.Insert(modelPresenca);
            }
            catch (Exception ex)
            {
                blnRetorno = false;
                throw ex;
            }
            finally
            {
                modelPresenca = null;
            }

            return blnRetorno;

        }
        public bool Insert(Model.Presenca oTipo)
        {

            bool blnRetorno = false;

            try
            {
                blnRetorno = repPresenca.Insert(oTipo);
            }
            catch (Exception ex)
            {
                blnRetorno = false;
                throw ex;
            }

            return blnRetorno;
        }

        public bool InsertLote(IList<Model.Presenca> lstFamilia)
        {

            bool blnRetorno = false;
            int intCont = 0;
            try
            {
                foreach (Model.Presenca oDado in lstFamilia)
                {
                    blnRetorno = Insert(oDado);
                    if (!blnRetorno)
                    {
                        throw new Exception(string.Format("Tentativa de Gravar Presença falhou. Antes foram gravados {0} de {1}  registros.", intCont.ToString() , lstFamilia.Count ));
                    }
                    intCont = intCont + 1;
                }
            }
            catch (Exception ex)
            {
                blnRetorno = false;
                throw ex;
            }

            return blnRetorno;
        }

        public IList<Model.Presenca> LoadAll()
        {

            IList<Model.Presenca> lstRetorno = null;

            try
            {
                lstRetorno = repPresenca.LoadAll();
            }
            catch (Exception ex)
            {
                lstRetorno = null;
                throw ex;
            }

            return lstRetorno;

        }

        public IList<Model.Familia> Load(int intReuniao)
        {

            IList<Model.Familia> lstRetorno = null;

            try
            {
                lstRetorno = repPresenca.Load(intReuniao);
            }
            catch (Exception ex)
            {
                lstRetorno = null;
                throw ex;
            }

            return lstRetorno;

        }

        public bool Update(Model.Presenca oTipo)
        {

            bool blnRetorno = false;

            try
            {
                blnRetorno = repPresenca.Update(oTipo);
            }
            catch (Exception ex)
            {
                blnRetorno = false;
                throw ex;
            }

            return blnRetorno;
        }

        public IList<Model.FamiliaPresenca> ObterPresencaPorMae(int intFamilia, int intAno)
        {

            IList<Model.FamiliaPresenca> lstRetorno = null;

            try
            {
                lstRetorno = repPresenca.ObterPresencaPorMae(intFamilia, intAno);
            }
            catch (Exception ex)
            {
                lstRetorno = null;
                throw ex;
            }

            return lstRetorno;

        }
    }
}
